File: firmware/lib/ead_core/src/ead/cobs.cpp
#include "ead/cobs.h"

namespace ead {

CobsWriter::CobsWriter(uint8_t* out, size_t cap) : out_(out), cap_(cap) { startBlock(); }

void CobsWriter::startBlock() {
  if (pos_ >= cap_) {
    ok_ = false;
    return;
  }
  codeIndex_ = pos_++;
  code_ = 1;
}

void CobsWriter::put(uint8_t byte) {
  if (!ok_) return;
  if (byte == 0) {
    out_[codeIndex_] = code_;
    startBlock();
    return;
  }
  if (pos_ >= cap_) {
    ok_ = false;
    return;
  }
  out_[pos_++] = byte;
  if (++code_ == 0xFF) {
    // A full block of 254 data bytes carries no implied zero.
    out_[codeIndex_] = code_;
    startBlock();
  }
}

void CobsWriter::put(const uint8_t* bytes, size_t n) {
  for (size_t i = 0; i < n; i++) put(bytes[i]);
}

size_t CobsWriter::finish() {
  if (!ok_) return 0;
  out_[codeIndex_] = code_;
  return pos_;
}

size_t cobsEncode(const uint8_t* in, size_t n, uint8_t* out, size_t cap) {
  CobsWriter w(out, cap);
  w.put(in, n);
  return w.finish();
}

size_t cobsDecode(const uint8_t* in, size_t n, uint8_t* out, size_t cap) {
  size_t i = 0;
  size_t o = 0;
  while (i < n) {
    const uint8_t code = in[i++];
    if (code == 0) return kCobsError;
    for (uint8_t k = 1; k < code; k++) {
      if (i >= n || in[i] == 0 || o >= cap) return kCobsError;
      out[o++] = in[i++];
    }
    if (code != 0xFF && i < n) {
      if (o >= cap) return kCobsError;
      out[o++] = 0;
    }
  }
  return o;
}

}  // namespace ead

File: dashboard/src-tauri/src/protocol/cobs.rs
//! Consistent Overhead Byte Stuffing (Cheshire & Baker), used to delimit
//! protocol messages on the USB byte stream.

use super::{ProtocolError, Result};

pub fn encode(data: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(data.len() + data.len() / 254 + 1);
    let mut code_index = 0;
    out.push(0); // placeholder for the first block's code
    let mut code: u8 = 1;
    for &byte in data {
        if byte == 0 {
            out[code_index] = code;
            code_index = out.len();
            out.push(0);
            code = 1;
            continue;
        }
        out.push(byte);
        code += 1;
        if code == 0xFF {
            // A full block of 254 data bytes carries no implied zero.
            out[code_index] = code;
            code_index = out.len();
            out.push(0);
            code = 1;
        }
    }
    out[code_index] = code;
    out
}

pub fn decode(data: &[u8]) -> Result<Vec<u8>> {
    let mut out = Vec::with_capacity(data.len());
    let mut i = 0;
    while i < data.len() {
        let code = data[i];
        if code == 0 {
            return Err(ProtocolError::BadCobs);
        }
        let block_end = i + code as usize;
        if block_end > data.len() {
            return Err(ProtocolError::BadCobs);
        }
        let block = &data[i + 1..block_end];
        if block.contains(&0) {
            return Err(ProtocolError::BadCobs);
        }
        out.extend_from_slice(block);
        i = block_end;
        if code != 0xFF && i < data.len() {
            out.push(0);
        }
    }
    Ok(out)
}
